sympy/combinatorics/homomorphisms.py

```diff
@@ -36,35 +36,57 @@ def _invs(self):
             if not (v in inverses
                     or v.is_identity):
                 inverses[v] = k
-        gens = image.strong_gens
+        if isinstance(self.codomain, PermutationGroup):
+            gens = image.strong_gens
+        else:
+            gens = image.generators
         for g in gens:
             if g in inverses or g.is_identity:
                 continue
             w = self.domain.identity
-            for s in image._strong_gens_slp[g]:
+            if isinstance(self.codomain, PermutationGroup):
+                parts = image._strong_gens_slp[g][:]
+                parts = [parts[len(parts)-i-1] for i in range(len(parts))]
+            else:
+                parts = g
+            for s in parts:
                 if s in inverses:
-                    w = inverses[s]*w
+                    w = w*inverses[s]
                 else:
-                    w = inverses[s**-1]**-1*w
+                    w = w*inverses[s**-1]**-1
             inverses[g] = w
+
         return inverses
 
     def invert(self, g):
         '''
-        Return an element of the preimage of `g`
+        Return an element of the preimage of `g`.
+        NOTE: If the codomain is an FpGroup, the inverse for equal
+        elements might not always be the same unless the FpGroup's
+        rewriting system is confluent. However, making a system
+        confluent can be time-consuming. If it's important, try
+        `self.codomain.make_confluent()` first.
 
         '''
-        if not isinstance(self.codomain, PermutationGroup):
-            raise NotImplementedError(
-                "Only elements of PermutationGroups can be inverted")
+        if isinstance(self.codomain, FpGroup):
+            g = self.codomain.reduce(g)
         if self._inverses is None:
             self._inverses = self._invs()
         image = self.image()
         w = self.domain.identity
-        for g in image.generator_product(g):
-            if g.is_identity:
+        if isinstance(self.codomain, PermutationGroup):
+            gens = image.generator_product(g)
+            gens = [gens[len(gens)-i-1] for i in range(len(gens))]
+        else:
+            gens = g
+        for i in range(len(gens)):
+            s = gens[i]
```

Comment:
I remembered why. Because `for s in gens:` would be equivalent to `for s in gens.array_form:` when `g` is a `FreeGroupElement`.  But when you call `gens` by index, the generator (or inverse) at position `i` is returned. It's actually a bit inconsistent. Maybe something needs to be done about it.